### DISCUSSION TITLE
Using Dagster Embedded ELT for data transfer and manipulation

### DISCUSSION THREAD

**Discussion by Codykilpatrick** [dagster-io/dagster#18837] [Q&A] [answered]
Hello friends! I am currently trying to use Dagster Embedded ELT to take data from mysql and insert it into a postgres database. I've got the connection stuff in order but I would like to also do some data transformation in the middle as the schemas don't match perfectly.

```
"""This module contains the embedded ELT code for the Forerunner data pipeline."""

from dagster import AssetSpec
from dagster_embedded_elt.sling import (
    SlingMode,
    SlingResource,
    SlingSourceConnection,
    SlingTargetConnection,
    build_sling_asset,
)

sling_resource = SlingResource(
    source_connection=SlingSourceConnection(
        type="mysql", connection_string="mysql://user:pass@localhost:3306/database"
    ),
    target_connection=SlingTargetConnection(
        type="postgres",
        connection_string="postgresql://user:pass@localhost:5432/database?sslmode=disable",
    ),
)

asset_def = build_sling_asset(
    asset_spec=AssetSpec("mysql_postgres_BTR"),
    source_stream="source.table",
    target_object="target.table",
    mode=SlingMode.INCREMENTAL,
    primary_key="id",
)
```

On example of what I am trying to do is I have a column that contains blob data and I want to decode it to a string before inserting it into the PostGres database. 

I've been really struggling to find documentation/examples so any help would be greatly appreciated! Also if this isn't the right tool for this job id be curious to hear about other options! Thanks for everything!

Specs:
Dagster: 1.5.13
dagster-embedded-elt: "^0.21.13"

**Comment by tacastillo** [comment]
Hi! Sling is focused on moving the data from source to destination and then doing transformations afterward. Would you be able to insert into a staging table in Postgres first, and then decode the blob when moving from the Postgres staging to your real destination table?

**Reply by Codykilpatrick** [reply]
Thats what I am thinking as a Plan B. I was going to first checkout if I could use a smart mysql query as the source stream that will convert the data on the way. But a staging table seems like the next option.

**Reply by PedramNavid** [reply, marked as answer]
Another option would be to provide a query instead of a table as a source stream that already performs the transformations.  https://docs.slingdata.io/sling-cli/run/database-to-database#database-database-custom-sql

Keep in mind complex queries may be more difficult to run successfully depending on database load.

**Reply by Codykilpatrick** [reply]
Thank you both for the options! I originally tried making a query but it turned into a bit of a nightmare so I reverted to trying to trying to dump it into a temporary table.

Now I am running into the issue of:
`pq: invalid byte sequence for encoding "UTF8": 0x00`
dagster embeded elt should support this correct? The errors are coming through weird and its hard to tell if this is the ETL or PostgreSQL not being happy.

**Reply by PedramNavid** [reply]
Looks like there’s an invalid null character that postgres doesn’t support in string columns: https://stackoverflow.com/questions/1347646/postgres-error-on-insert-error-invalid-byte-sequence-for-encoding-utf8-0x0

One option might be to write data as parquet instead of directly into postgres, then use something like duckdb to do transforms/cleanup  and then write that back into postgres using duckdb or sling.

**Reply by Codykilpatrick** [reply]
Hey thank you guys for all of the help sorry I was slow with the replies. I did end up using the embedded-elt to go from mySql to postgres staging table. (converting blobs to hex strings along the way). Then used postgres function to do the final transformation and move to the final table so it all worked out!

**Reply by PedramNavid** [reply]
Great to hear it! Good luck on your journey!
